### ISSUE TITLE
Rule Change: `beforeStatementContinuationChars`(in rule `semi`) does not work as expected

### ISSUE THREAD

**Issue by andjsrk** [eslint/eslint#19349]
### What rule do you want to change?

semi

### What change do you want to make?

Generate more warnings

### How do you think the change should be implemented?

A new option

### Example code

```js
/* eslint @stylistic/semi: ["error", "never", { "beforeStatementContinuationChars": "always" }] */

// 1)
do {} while (true)
[1, 2, 3].forEach()

// 2)
function foo() {}
[1, 2, 3].forEach()
foo()
[1, 2, 3].forEach()
const bar = foo()
[1, 2, 3].forEach()
// etc
```

### What does the rule currently do for this code?

`1)`: Currently reports missing semicolon. (expected)
`2)`: Currently reports nothing. (unexpected)
Note that `2)` is the same as the original ESLint behavior (https://github.com/eslint/eslint/issues/11053#issuecomment-435658719), but it *is not documented* (the document simply says "`"beforeStatementContinuationChars": "always"` requires semicolons at the end of statements if the next line starts with `[`, `(`, `/`, `+`, or `-`.").

### What will the rule do after it's changed?

The rule reports the every line that starts a statement with a token that may cause ASI hazards (as the document says).

### Participation

- [ ] I am willing to submit a pull request to implement this change.

### Additional comments

I'm actually using rule `semi-style` with option `"first"` so the code with semicolons looks like:
```js
do {} while (true)
;[1, 2, 3].forEach()

function foo() {}
;[1, 2, 3].forEach()

foo()
;[1, 2, 3].forEach()
```
So in my case the configuration `"never", { "beforeStatementContinuationChars": "always" }` keeps consistency for these ASI hazard exceptions so let it improve portability(IDK this is an appropriate word though) of the statement.
I want to keep this consistency regardless of the existence of ESLint rule [`no-unexpected-multiline`](https://eslint.org/docs/latest/rules/no-unexpected-multiline).

**Comment by andjsrk** [comment]
Misposted; closing the issue.
